python/ambassador/envoy/v3/v3listener.py

```diff
@@ -71,41 +74,65 @@ def add_tcphost(self, tcpmapping: IRTCPMappingGroup) -> None:
             )
 
         if len(self.hosts) > 0:
-            # If we have SNI, then each host gets its own chain, so we should never have more than 1
-            # self.hosts; if we don't have SNI then a single TLSContext takes over the entire chain
-            # and so we shouldn't have more than 1 self.hosts then either.
+            # If we have SNI, then each tcp host gets its own Filter Chain, so we should never have more than 1
+            # entry in self.hosts; if we don't have SNI then a single FilterChain with no filter_chain_match
+            # takes over the entire chain and so we still should not have more than 1 self.hosts then either.
             other = next(iter(self.hosts.values()))
             other_type = "TCPMapping" if isinstance(other, IRTCPMappingGroup) else "Host"
             tcpmapping.post_error(
-                "TCPMapping {tcpmapping.name}: discarding because it conflicts with {other_type} {other.name}"
+                f"TCPMapping {tcpmapping.name}: discarding because it conflicts with {other_type} {other.name}"
```

Comment:
>so we're not going to allow a Host resource with hostname example.com:5000 and a TCPMapping with host example.com regardless of port specified on the TCPMapping?

Correct, in `v3listeners.finalize` we process TCPMappings first and their filter chains get precedence on a Listener. Luke left a comment in that function that outlines it well, the reasoning for this has to do with providing sensible upgrade paths. 

Since TCPMapping gets precedence in the scenario you outline they are both trying to generate a FilterChain with the same FilterChainMatch (TLS & SNI (`example.com`) which Envoy doesn't allow. Technically, you could differentiate these filter chains by adding alpn `application_protocol: [h2,http/1.1]` and Envoy would accept it but Envoy itself calls out the fact that ALPN is not reliable so I think this would cause more headaches then it benefits. 

>What if you specified a different port on the TCPMapping? Theoretically you should be able to put them on different filter chains with different filter chain matching based on SNI & port.
>
>Or this that case handled elsewhere since believe a new listener is created if they aren't any with the specified TCPMapping port

Correct, that would work because the TCPMapping would attach to an existing Listener or implicitly create one behind the scenes for that `address/port` thus there would be no overlap.

## Couple of things

1. I think this function where you commented on, can be confusing because it relies on having knowledge of the TCPMappings taking precedence in `V3Listeners.finalize` and also the method chaining where  the `V3Listener.add_chain` is shared by TCP/HTTP chains and then returns the V3Chain. So, mentally you have switch context pending which code path led you here.

2. While I was looking into this I did find a bug when trying to apply a non-tls cleartext TCPMapping to a listener that serves HTTP. I have captured that here: https://github.com/emissary-ingress/emissary/issues/4825